optuna/integration/sklearn.py

```diff
@@ -237,10 +306,18 @@ def __call__(self, trial: Trial) -> float:
                 return_train_score=self.return_train_score,
                 scoring=self.scoring,
             )
+            self.multimetric_ = "test_score" not in scores
+
+            # check refit_metric now for a callabe scorer that is multimetric
+            if callable(self.scoring) and self.multimetric_:
+                self._check_refit_for_multimetric(scores)
 
         self._store_scores(trial, scores)
 
-        return trial.user_attrs["mean_test_score"]
+        if self.multimetric_:
+            return trial.user_attrs[f"mean_test_{self.refit}"]
```

Comment:
Does this work well when `refit` is callable?

Reply:
If it is multi-metric,  the `refit` must be `str`, which is the scikit-learn implementation, as well as documented in our implementation.

Reply:
I think the scikit-learn's documentation looks subtle.  According to this sklearn, refit can be callable for both cases.

https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/model_selection/_search.py#L856-L859

Indeed, the following code works.

```python
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform
import numpy as np
iris = load_iris()

logistic = LogisticRegression(solver='saga', tol=1e-2, max_iter=200,
                              random_state=0)
distributions = dict(C=uniform(loc=0, scale=4),
                     penalty=['l2', 'l1'])
def refit_eval(x):
    return np.argmin(x["mean_test_neg_mean_squared_error"])

clf = RandomizedSearchCV(logistic, distributions, random_state=0, scoring=['neg_mean_squared_error', 'accuracy'],
                         refit=refit_eval)
search = clf.fit(iris.data, iris.target)
search.best_params_
```

Reply:
It is interesting that your code indeed runs without any errors...
For me, it looks like a kind of undefined behaviors, and this is scikit-learn's issue.
[Their docs](https://github.com/scikit-learn/scikit-learn/blob/28ef9973362257e0627bd39db9b788c93f49362f/sklearn/model_selection/_search.py#L1474-L1476) states that:
```
For multiple metric evaluation, this needs to be a `str` denoting the
scorer that would be used to find the best parameters for refitting
the estimator at the end.
```

Reply:
Let's ask scikit-learn community: https://github.com/scikit-learn/scikit-learn/discussions/21095

Reply:
Thx!

Reply:
According to the maintainer, it is not appropriate usage of `refit`. I think the compatibility to scikit-learn is important but this is a subtle situation. I suggest not changing the codes and leaving a comment that mentions we do not support callable `refit` (yet). What do you think?
https://github.com/scikit-learn/scikit-learn/discussions/21095#discussioncomment-1379615